### PR TITLE
Add cubic and quadratic Bézier curves to Workplane and Sketch

### PR DESCRIPTION
**Pull request by dov**
This change adds quadratic and cubic Bézier curves to the Workplane and the Sketch classes

### CODE REVIEW

**Comment by neri-engineering** [comment]
I had a quick look through the changes and these look to be thorough changes.  Unfortunately I cannot vouch for nor can I condemn these changes b/c I have not mastered the art of Git or the art of running CQ from source code (it's on my TODO list after I finish with elliptical B-splines).

But here are my thoughts upon seeing the changes (besides the test code).

- In `cq.py`, that `quadraticCurveTo()` and `cubicCurveTo()` accept 2D coordinates is consistent with `lineTo()` and `moveTo()`.
- In `shapes.py`, the methods added take `VectorLike` objects, which can either be 2D or 3D.  So from the API perspective there is support for specifying 3D coordinates.  Great.
- In `sketch.py`, `Point` datatype includes `Vector`, which has ability to specify in 3D.

I just hope that the handing of 2D and 3D works.  The datatypes used support various formats, they use the `Union` construct typically.  I can't say anything more intelligent because I'm not yet an expert w.r.t. CQ and Python.

**Comment by adam-urbanczyk** [comment]
Hey, thanks. I'm personally not convinced that this is needed, but OK. If you want to get it merged, I'll ask you to change the API so that it follows what spline does. Also please support any bezier, not just cubic/quadratic. The API should look approximately like this:
```python
cq.Shape.makeBezier
cq.Workplane.bezier(
    self: T,
    listOfXYTuple: Iterable[VectorLike],
    forConstruction: bool = False,
    includeCurrent: bool = False,
    makeWire: bool = False,
)
cq.Sketch.bezier(
    self: T,
    pts: Iterable[Point],
    tag: Optional[str] = None,
    forConstruction: bool = False,
)
```
Rationale: generality, consistency with spline and 3D support.

**Comment by dov** [comment]
@adam-urbanczyk Thanks. A question, with the above change and if someone wants to draw a 2D bezier curve e.g. in the XZ plane, do the vector coordinates need to provided with x,y and z coordinates, with the ycoordinate==0?

**Comment by adam-urbanczyk** [comment]
I'm not sure what you exactly mean. You can provide 2D coordinates in the local coordinate system, see: https://cadquery.readthedocs.io/en/latest/examples.html#defining-an-edge-with-a-spline.

**Comment by dov** [comment]
Ok. Thanks. That's what I thought. But for whatever reason, after I tried implementing your suggestions, I could only get a solid after extrusion if doing chosing an XY Workplane. If choosing an XZ WorkPlane, I got a degenerate solid:

```python
import sys
sys.path.insert(0,'/path/to/dev/cadquery/')
import cadquery as cq

for plane in ('XY','XZ'):
  res = (cq
         .Workplane(plane)
         .bezier([(0,0), (1, 2), (5, 0)])
         .bezier([(5,0), (1, -2), (0, 0)])
         .close()
         .extrude(1)
         )
  
  print(f'{plane=} Volume={res.findSolid().Volume()}')
```

outputs:
```
plane='XY' Volume=6.666666666666666
plane='XZ' Volume=0.0
```

The implementation in `Shape.py` is quite straight forward:
```
    @classmethod
    def makeBezier(
        cls,
        points: List[Vector]
    ) -> "Edge":
        """
        Create a cubic Bézier Curve from the points.

        :param listOfVector: a list of Vectors that represent the points.
            The edge will pass through the first and the last point,
            and the inner points are Bézier control points.
        :return: An edge
        """

        # Convert to a TColgp_Array1OfPnt
        arr = TColgp_Array1OfPnt(1, len(points))
        for i, v in enumerate(points):
            arr.SetValue(i + 1, Vector(v).toPnt())
      
        bez = Geom_BezierCurve(arr)

        return cls(BRepBuilderAPI_MakeEdge(bez).Edge())
```

and in `cq.py`:

```python
    def bezier(
        self: T,
        listOfXYTuple: Iterable[VectorLike],
        forConstruction: bool = False,
        includeCurrent: bool = False,
        makeWire: bool = False
      
    ) -> T:
        if includeCurrent:
            listOfXYTuple = [self._findFromPoint(False)] + [v for v in listOfXYTuple]

        e = Edge.makeBezier(listOfXYTuple)

        if makeWire:
            rv_w = Wire.assembleEdges([e])
            if not forConstruction:
                self._addPendingWire(rv_w)
        elif not forConstruction:
            self._addPendingEdge(e)

        return self.newObject([rv_w if makeWire else e])
```

Do you have any idea, why the local coordinate system isn't "catching"?

**Comment by adam-urbanczyk** [comment]
https://github.com/CadQuery/cadquery/blob/153ed3f667911e909ba8b93bddd297defc7cd42f/cadquery/cq.py#L1887

**Comment by dov** [comment]
> https://github.com/CadQuery/cadquery/blob/153ed3f667911e909ba8b93bddd297defc7cd42f/cadquery/cq.py#L1887

Great! Thanks. That solved the problem.

**Comment by dov** [comment]
@adam-urbanczyk One thing that is bothering me with the suggested API for `Sketch` is that there is no support for starting from the current point. Would it be ok if i add a `includeCurrent` flag for the Sketch API as well so that instead of writing:

```python
s1 = (
  cq.Sketch()
  .segment((0,0), (0,0.5))
  .bezier(((0, 0.5), (-1, 2), (1, 0.5), (5, 0)))
  .bezier(((5,0), (1, -0.5), (-1, -2), (0, -0.5)))
  .close()
```

I can write:
```python
s1 = (
  cq.Sketch()
  .segment((0,0), (0,0.5))
  .bezier(((-1, 2), (1, 0.5), (5, 0)), includeCurrent=True)
  .bezier(((1, -0.5), (-1, -2), (0, -0.5)), includeCurrent=True)
  .close()
  )
```

**Comment by adam-urbanczyk** [comment]
Sorry, but that does not mesh with the current sketch API.

**Comment by dov** [comment]
Ok. As you wish. But in the future perhaps you can add some place holder like `cq.CP`, `cq.current_pos`, or `cq.cont` as a shortcut for the current point? That would benifit your `spline()` method as well.

```
s1 = (
  cq.Sketch()
  .segment((0,0), (0,0.5))
  .bezier((cq.CP, (0, 0.5), (-1, 2), (1, 0.5), (5, 0)))
  .bezier((cq.CP, (5,0), (1, -0.5), (-1, -2), (0, -0.5)))
  .close()
```

or even:
```
_ = cq.CP
s1 = (
  cq.Sketch()
  .segment((0,0), (0,0.5))
  .bezier((_, (-1, 2), (1, 0.5), (5, 0)))
  .bezier((_, (1, -0.5), (-1, -2), (0, -0.5)))
  .close()
```

In any case, that is outside the scope of this pull request.

**Comment by adam-urbanczyk** [comment]
@dov are you going to fix the coverage of your PR?

**Comment by adam-urbanczyk** [comment]
Preferably 100% patch coverage

**Comment by jmwright** [comment]
@dov It looks like the black formatting check is falling in AppVeyor. Two links are provided below in case you haven't used black before.

[Install black](https://github.com/CadQuery/cadquery/blob/master/environment.yml#L31)

[Use black](https://github.com/CadQuery/cadquery/blob/master/appveyor.yml#L35)

**Comment by dov** [comment]
> @dov It looks like the black formatting check is falling in AppVeyor. Two links are provided below in case you haven't used black before.
> 
> [Install black](https://github.com/CadQuery/cadquery/blob/master/environment.yml#L31)
> 
> [Use black](https://github.com/CadQuery/cadquery/blob/master/appveyor.yml#L35)

Thanks @jmwright . I'm on to it one iteration at a time.

**Comment by jmwright** [comment]
@adam-urbanczyk All green now.

**Review by adam-urbanczyk — CHANGES_REQUESTED** [review]
LGTM, I removed one unused variable. Could you clarify or remove the comment you made in tests?

File: tests/test_workplanes.py [hunk 1]
```diff
@@ -250,3 +250,53 @@ def test_mirror_face(self):
...
+        r = (
+            Workplane("XZ")
+            .bezier([(0, 0), (1, 2), (5, 0)])
+            .bezier([(1, -2), (0, 0)], includeCurrent=True)
+            .close()
+            .extrude(1)
+        )
+
+        bbBox = r.findSolid().BoundingBox()
+        # Why is the bounding box larger than expected?
```
[the remark this answers is not in the document]
Comment (dov) [reply tests/test_workplanes.py]: Ok, thanks. I don't know what was the issue. It is ok now. I just forced pushed an update.

**Review by adam-urbanczyk — APPROVED** [review]

File: cadquery/occ_impl/shapes.py [hunk 2]
```diff
@@ -2091,6 +2092,26 @@ def makeLine(cls, v1: VectorLike, v2: VectorLike) -> "Edge":
             BRepBuilderAPI_MakeEdge(Vector(v1).toPnt(), Vector(v2).toPnt()).Edge()
         )
 
+    @classmethod
+    def makeBezier(cls, points: List[Vector]) -> "Edge":
+        """
+        Create a cubic Bézier Curve from the points.
+
+        :param listOfVector: a list of Vectors that represent the points.
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:2100]: Docstring param name mismatch. 
The docs list both points, and listOfVector:
https://cadquery--1529.org.readthedocs.build/en/1529/classreference.html

File: tests/test_workplanes.py [hunk 3]
```diff
@@ -250,3 +250,53 @@ def test_mirror_face(self):
...
+        self.assertTrue(len(r.ctx.pendingEdges) == 1)
+        r = (
+            r.lineTo(5, -0.1)
+            .bezier([(2, -3), (1, 0), (0, 0)], includeCurrent=True)
+            .close()
+            .extrude(1)
+        )
+
+        bbBox = r.findSolid().BoundingBox()
+        print(bbBox)
```
Comment (jmwright) [inline tests/test_workplanes.py:279]: Is this a stray print statement?

**Review by jmwright — APPROVED** [review]
Other than the one comment from @lorenzncode and the other from me, it looks good. Thanks @dov . Have you updated your gist to work with this latest changes in this PR yet?

File: cadquery/occ_impl/shapes.py (hunk 2 above)
Comment (dov) [reply cadquery/occ_impl/shapes.py:2100]: Fixed! Thanks!

File: tests/test_workplanes.py (hunk 3 above)
Comment (dov) [reply tests/test_workplanes.py:279]: Fixed! Thanks!

**Comment by dov** [comment]
> Other than the one comment from @lorenzncode and the other from me, it looks good. Thanks @dov . Have you updated your gist to work with this latest changes in this PR yet?

I assume you mean the svg path to cadquery gist. I just updated it. Thanks for reminding me.

**Comment by adam-urbanczyk** [comment]
Thanks @dov !

**Pull request merged**